v2/yarpcprotobuf/protoc-gen-yarpc-go/internal/generator/registry.go

```diff
@@ -0,0 +1,293 @@
+
+	/* Plugin parameters */
+	packages map[string]string
+}
+
+func newRegistry(req *plugin.CodeGeneratorRequest) (*registry, error) {
+	r := &registry{
+		files:    make(map[string]*File),
+		messages: make(map[string]*Message),
+		packages: make(map[string]string),
+		imports: Imports{
+			"context":                            "context",
+			"go.uber.org/fx":                     "fx",
+			"github.com/gogo/protobuf/proto":     "proto",
+			"go.uber.org/yarpc/v2":               "yarpc",
+			"go.uber.org/yarpc/v2/yarpcprotobuf": "yarpcprotobuf",
+		},
+	}
+	// Process command-line plugin parameters.
+	if req.Parameter != nil {
+		for _, p := range strings.Split(req.GetParameter(), ",") {
+			param := strings.SplitN(p, "=", 2)
+			if len(param) != 2 {
+				return nil, fmt.Errorf("package modifiers should include a single '='")
+			}
+			flag, value := param[0], param[1]
+			switch {
+			case strings.HasPrefix(flag, "M"):
+				r.packages[flag[1:]] = value
+			default:
+				return nil, fmt.Errorf("unknown flag: %v", flag)
+			}
+		}
+	}
+	return r, r.Load(req)
+}
+
+// Load registers all of the Proto types provided in the
+// CodeGeneratorRequest with the registry.
+func (r *registry) Load(req *plugin.CodeGeneratorRequest) error {
+	for _, f := range req.GetProtoFile() {
+		r.loadFile(f)
+	}
+	for _, name := range req.FileToGenerate {
+		target, ok := r.files[name]
+		if !ok {
+			return fmt.Errorf("file target %q was not registered", name)
+		}
+		for _, s := range target.descriptor.GetService() {
+			if err := r.loadService(target, s); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetData returns the template data the corresponds
+// to the given filename.
+func (r *registry) GetData(filename string) (*Data, error) {
+	f, err := r.getFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	return &Data{
+		File:    f,
+		Imports: r.imports,
+	}, nil
+}
+
+// getFile returns the File that corresponds to the
+// given filename.
+func (r *registry) getFile(filename string) (*File, error) {
+	f, ok := r.files[filename]
+	if !ok {
+		return nil, fmt.Errorf("file %q was not found", filename)
+	}
+	return f, nil
+}
+
+// getMessage returns the Message that corresponds to the
+// given name. This method expects the input to be formed
+// as an input or output type, such as .foo.Bar.
+func (r *registry) getMessage(name string) (*Message, error) {
+	// All input and output types are represented as
+	// .$(Package).$(Message), so we explicitly trim
+	// the leading '.' prefix.
+	msg := strings.TrimPrefix(name, ".")
+	m, ok := r.messages[msg]
+	if !ok {
+		return nil, fmt.Errorf("message %q was not found", msg)
+	}
+	return m, nil
+}
+
+// loadFile registers the given file's message types.
+// Note that we load the messages for all files up-front so that
+// all of the message types potentially referneced in the proto
+// services can reference these types.
+func (r *registry) loadFile(f *descriptor.FileDescriptorProto) {
```

Comment:
Not sure if this is even possible, but should we do a nil check here first in case `f == nil`?
If it is, we'd load an empty `File` to the `files` map

Reply:
This shouldn't be necessary given that we are using the supported accessors, such as `f.Get*()`. All of the accessors are are nil-safe, e.g.
```
func (m *FileDescriptorProto) GetName() string {
  if m != nil && m.Name != nil {
    return *m.Name
  }
  return ""
}
```

Also note that this would end up adding a lot of uncovered lines of code in a variety of functions, such as `loadService` and `loadMessage`, so I'd rather not add this in for the time being. 

What are your thoughts?